### PR TITLE
Modernize documentation

### PR DESCRIPTION
**Pull request by SzymonKaminski**
This PR aims to modernize the bundle's documentation by resolving several opened issues.

- All URLs found in docs were checked for correctness and updated accordingly, new links were added where they could be helpful for newcomers (Form, Security, environment variables) - #538 

- Code examples no longer refer to Acme bundles; usage of Symfony Flex is recommended, configuration steps for projects not using Flex were updated for completeness - #536 

- code examples comply with Doctrine coding standard with few exceptions mentioned below  - #544

- some sentences were reworded to avoid belittling words - #539  and https://symfony.com/doc/current/contributing/documentation/standards.html

- a tip mentioning Replica Set option in connection string was added with link to relevant MongoDB docs page - #374 

- typos were fixed and unclear paragraphs were reworded

One twig template file still uses code-block `html+jinja` as `html+twig` would not work currently as mentioned in https://github.com/doctrine/DoctrineMongoDBBundle/pull/613#pullrequestreview-355635224

`phpcs.xml.dist` file from this repository was used when running phpcs in SF skeleton, therefore parameter and return type hints were not added. Additionally `@MongoDB` annotations spanning multiple lines (instead of one) were left unchanged. 

Every issue is resolved in separate commit containing brief description of changes, unless sensible splitting of given line(s) was hard to achieve.

Using doctrine/rst-parser directly in my Symfony skeleton or SenseException/rst-builder produced errors as mentioned in https://github.com/doctrine/DoctrineMongoDBBundle/pull/613#issuecomment-584828215 (e.g. Unknown directive: "tip" for line ".. tip::").
I set up doctrine-website locally and confirmed that docs based on this PR can be built without any errors.

### CODE REVIEW

**Comment by SenseException** [comment]
`html+twig` will currently run into a warning.

**Comment by malarzm** [comment]
@SzymonKaminski thank you a lot for your effort and this PR! Just wanted you to know that we need to handle #613 first, then this PR will prolly need a rebase

**Comment by malarzm** [comment]
@SzymonKaminski sorry for the long time, but we've finally merged #613 and merged it up into `4.1.x` branch. Could you please rebase onto that branch and also target `4.1.x` with this PR? That's current maintained branch so docs changes you want to introduce need to go there. Also I hope that work done in #613 and in #622 won't make too much trouble for you but I wanted to let you start with green build :)

**Comment by SzymonKaminski** [comment]
The lint action is green and I believe I haven't missed anything during the rebase :tada:

File: Resources/doc/first_steps.rst [hunk 1]
```diff
@@ -38,9 +38,9 @@ inside the ``Document`` directory of your ``AcmeStoreBundle``:
     }
 
 The class - often called a "document", meaning *a basic class that holds data* -
-is simple and helps fulfill the business requirement of needing products
-in your application. This class can't be persisted to Doctrine MongoDB yet -
-it's just a simple PHP class.
+helps fulfill the business requirement of needing products in your application.
+This class can't be persisted to Doctrine MongoDB yet - currently it's
+only plain PHP class.
```
Comment (malarzm) [inline Resources/doc/first_steps.rst]:
```suggestion
only a plain PHP class.
```

File: Resources/doc/first_steps.rst [hunk 2]
```diff
@@ -51,8 +51,8 @@ you to persist entire *objects* to MongoDB and fetch entire objects out of
 MongoDB. This works by mapping a PHP class and its properties to entries
 of a MongoDB collection.
 
-For Doctrine to be able to do this, you just have to create "metadata", or
-configuration that tells Doctrine exactly how the ``Product`` class and its
+For Doctrine to be able to do this, you have to create "metadata", or
+configuration, that tells Doctrine exactly how the ``Product`` class and its
```
Comment (malarzm) [inline Resources/doc/first_steps.rst]: There should be no `,`
```suggestion
configuration that tells Doctrine exactly how the ``Product`` class and its
```

File: Resources/doc/first_steps.rst [hunk 3]
```diff
@@ -115,9 +115,8 @@ Persisting Objects to MongoDB
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 Now that you have a mapped ``Product`` document complete with getter and
-setter methods, you're ready to persist data to MongoDB. From inside a controller,
-this is pretty easy. Add the following method to the ``DefaultController``
-of the bundle:
+setter methods, you're ready to persist data to MongoDB. Let's try it from inside
+a controller. Create new Controller class inside source directory of your project:
```
Comment (malarzm) [inline Resources/doc/first_steps.rst]: ``DefaultController` was already created in this chapter, please bring back the part about adding a new method to it

File: Resources/doc/first_steps.rst [hunk 4]
```diff
@@ -321,16 +320,16 @@ method of the document manager:
...
+however, isn't executed until the ``flush()`` method is called.
 
 Querying for Objects
 --------------------
 
 As you saw above, the built-in repository class allows you to query for one
-or many objects based on an number of different parameters. When this is
-enough, this is the easiest way to query for documents. Of course, you can
+or many objects based on any number of different parameters. When this is
+enough, this is the recommended way to query for documents. Of course, you can
```
Comment (malarzm) [inline Resources/doc/first_steps.rst]: This is literally the easiest way, no need to avoid this word altogether. Also personally I'm far from recommending it :)

File: Resources/doc/events.rst [hunk 5]
```diff
@@ -64,11 +64,11 @@ Event Subscribers
 -----------------
 
 Use the ``doctrine_mongodb.odm.event_subscriber`` tag
-to register a subscriber. Subscribers are responsible for implementing
-``Doctrine\Common\EventSubscriber`` and a method for returning the events
-they will observe. For this reason, this tag has no ``event`` attribute;
-however, the ``connection``, ``priority`` and ``lazy`` attributes are
-available.
+to register a subscriber. Subscribers must implement interface
```
Comment (malarzm) [inline Resources/doc/events.rst]: I think this should read "must implement the `...` interface

File: Resources/doc/installation.rst [hunk 6]
```diff
@@ -11,8 +11,8 @@ in the `installation chapter`_ of the Composer documentation.
 Install the bundle with Symfony Flex
 ------------------------------------
 
-A Flex recipe for the DoctrineMongoDBBundle is provided through a Contrib Recipe
-therefore you need to allow its usage:
+A Flex recipe for the DoctrineMongoDBBundle is provided as Contrib Recipe.
```
Comment (malarzm) [inline Resources/doc/installation.rst]:
```suggestion
A Flex recipe for the DoctrineMongoDBBundle is provided as a Contrib Recipe.
```

File: Resources/doc/installation.rst [hunk 7]
```diff
@@ -82,16 +82,16 @@ the MongoDB ODM across your application:
 Authentication
 --------------
 
-If you use authentication on your MongoDB database you can the provide username, 
-password, and authentication database in the following way:
+If you use authentication on your MongoDB database you can then provide username,
```
Comment (malarzm) [inline Resources/doc/installation.rst]: I'd remove `then`

File: Resources/doc/installation.rst [hunk 8]
```diff
@@ -82,16 +82,16 @@ the MongoDB ODM across your application:
...
+password and authentication database in the following way:
```
Comment (malarzm) [inline Resources/doc/installation.rst]: let's add oxford comma while we're here
```suggestion
password, and authentication database in the following way:
```

File: Resources/doc/first_steps.rst [hunk 9]
```diff
@@ -5,16 +5,7 @@ The best way to understand the Doctrine MongoDB ODM is to see it in action.
...
-    If you want to follow along with the example in this chapter, create
-    an ``AcmeStoreBundle`` via:
-
-    .. code-block:: bash
-
-        php bin/console generate:bundle --namespace=Acme/StoreBundle
-
-A Simple Example: A Product
+An Introductory Example: A Product
 ---------------------------
```
Comment (malarzm) [inline Resources/doc/first_steps.rst]: Please add some `-` so the line's length matches the one above

File: Resources/doc/config.rst [hunk 10]
```diff
@@ -133,7 +133,7 @@ can control. The following configuration options exist for a mapping:
   this path is relative it is assumed to be relative to the bundle root. This
   only works if the name of your mapping is a bundle name. If you want to use
   this option to specify absolute paths you should prefix the path with the
-  kernel parameters that exist in the DIC (for example %kernel.root_dir%).
+  kernel parameters that exist in the DIC (for example %kernel.project_dir%).
```
Comment (malarzm) [inline Resources/doc/config.rst]: Should we put %...% in backticks so it looks like a code?

File: Resources/doc/installation.rst [hunk 11]
```diff
@@ -31,38 +31,36 @@ To install DoctrineMongoDBBundle with Composer run the following command:
 
     composer require doctrine/mongodb-odm-bundle
 
-All that is left to do is to update your ``AppKernel.php`` file, and
-register the new bundle:
+
+Enable the Bundle
```
Comment (malarzm) [inline Resources/doc/installation.rst]: Original intention was that this text belongs to `Install the bundle with Composer` chapter as the bundle will be automatically enabled when using Flex.

File: Resources/doc/installation.rst [hunk 12]
```diff
@@ -31,38 +31,36 @@ To install DoctrineMongoDBBundle with Composer run the following command:
...
-    }
+        Doctrine\Bundle\MongoDBBundle\DoctrineMongoDBBundle::class => ['all' => true],
+    ];
 
 Configuration
 -------------
 
-To get started, you'll need some basic configuration that sets up the document
-manager. The easiest way is to enable ``auto_mapping``, which will activate
+Flex recipe will automatically create ``config/packages/doctrine_mongodb.yaml``
```
Comment (malarzm) [inline Resources/doc/installation.rst]:
```suggestion
Flex recipe will automatically create the ``config/packages/doctrine_mongodb.yaml``
```

File: Resources/doc/installation.rst [hunk 13]
```diff
@@ -31,38 +31,36 @@ To install DoctrineMongoDBBundle with Composer run the following command:
...
+file with default configuration. Without Flex you need to create the file
+manually and fill it with some basic configuration that sets up the document manager.
+The recommended way is to enable ``auto_mapping``, which will activate
 the MongoDB ODM across your application:
 
 .. code-block:: yaml
 
-    # app/config/parameters.yml
-    parameters:
-        mongodb_server: "mongodb://localhost:27017"
```
Comment (malarzm) [inline Resources/doc/installation.rst]: why is this part removed without replacement? We're using `mongodb_server` parameter in next example

File: Resources/doc/first_steps.rst (hunk 3 above)
Comment (SzymonKaminski) [reply Resources/doc/first_steps.rst]: At the top of the file I've removed section with `bin/console generate:bundle --namespace=Acme/StoreBundle` command from which DefaultController came from

File: Resources/doc/config.rst [hunk 14]
```diff
@@ -51,20 +51,20 @@ Sample Configuration
 .. tip::
 
     If each environment requires a different MongoDB connection URI, you can
-    define it in a separate parameter and reference it in the bundle config:
+    `define it as environment variable`_ and reference it in the bundle config:
```
Comment (malarzm) [inline Resources/doc/config.rst]:
```suggestion
    `define it as an environment variable`_ and reference it in the bundle config:
```

File: Resources/doc/installation.rst (hunk 13 above)
Comment (malarzm) [reply Resources/doc/installation.rst]: Seems like it's making a come back later with env variables, nvm me 👍

Comment (malarzm) [reply Resources/doc/installation.rst]: Or not, it's in different file, my questions still stands considering the scope of installation

File: Resources/doc/config.rst [hunk 15]
```diff
@@ -363,7 +363,7 @@ Connecting to a pool of mongodb servers on 1 connection
 
 It is possible to connect to several mongodb servers on one connection if
 you are using a replica set by listing all of the servers within the connection
-string as a comma separated list.
+string as a comma separated list and using ``replicaSet`` option.
```
Comment (malarzm) [inline Resources/doc/config.rst]: "and" made me think the options is something more
```suggestion
string as a comma separated list provided with the ``replicaSet`` option.
```

File: Resources/doc/config.rst [hunk 16]
```diff
@@ -373,7 +373,7 @@ string as a comma separated list.
             # ...
             connections:
                 default:
-                    server: "mongodb://mongodb-01:27017,mongodb-02:27017,mongodb-03:27017"
+                    server: "mongodb://mongodb-01:27017,mongodb-02:27017,mongodb-03:27017/?replicaSet=replSetName"
```
Comment (malarzm) [inline Resources/doc/config.rst]: @alcaeus is this necessary?

File: Resources/doc/first_steps.rst (hunk 3 above)
Comment (malarzm) [reply Resources/doc/first_steps.rst]: Ah sorry then 👍  so we need to just change
```suggestion
a controller. Create a new Controller class inside source directory of your project:
```

Also what do you think about having entire controller in example? Showing only method to add made sense when controller was already there, now it looks a bit off IMO

File: Resources/doc/first_steps.rst (hunk 4 above)
Comment (SzymonKaminski) [reply Resources/doc/first_steps.rst]: Right, no need to overdo it in this case

File: Resources/doc/installation.rst (hunk 13 above)
Comment (SzymonKaminski) [reply Resources/doc/installation.rst]: Added it back, inside `config/services.yaml`

File: Resources/doc/first_steps.rst (hunk 3 above)
Comment (SzymonKaminski) [reply Resources/doc/first_steps.rst]: Ok, will add class name to the example. Also the file path inside codeblock in the next section _Fetching Objects from MongoDB_. That example doesn't need full class, right?

File: Resources/doc/installation.rst (hunk 11 above)
Comment (SzymonKaminski) [reply Resources/doc/installation.rst]: Can this section be dropped entirely then? I left it only for completeness and added sentence `Your bundle should be automatically enabled if you use Flex`, but I agree it doesn't provide much value.

File: Resources/doc/first_steps.rst (hunk 3 above)
Comment (malarzm) [reply Resources/doc/first_steps.rst]: > That example doesn't need full class, right?

Right :+1:

File: Resources/doc/installation.rst (hunk 11 above)
Comment (malarzm) [reply Resources/doc/installation.rst]: I'm not entirely sure what I had in mind 2 weeks ago, on a fresh look this looks OK. So let's ignore my original comment :)

File: Resources/doc/config.rst (hunk 16 above)
Comment (alcaeus) [reply Resources/doc/config.rst]: Yes, if you are connecting to a replica set. See the corresponding [documentation](https://docs.mongodb.com/manual/reference/connection-string/#replica-set-option).

**Comment by malarzm** [comment]
On a weekend thought: this is PR is so good I'm gonna change that one thing off after merging, let's not keep this open anymore. Thanks a lot @SzymonKaminski! 🎉
